src/outcomes.py

```diff
@@ -1,39 +1,127 @@
-from typing import List, Tuple
+from typing import List, Tuple, Union
 import numpy as np
+import pandas as pd
 
+UnionArray = Union[np.ndarray, pd.DataFrame]
 
-def plurality_outcome(s: np.array) -> List[Tuple[str, int]]:
-    first_prefs = s[0, :]
+
+# Class to store the results of the voting,
+# like a dictionary but with a winner field
+class Result(dict):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.winner = self.get_winner()
+
+    # Method to get the winner of the voting, resolving ties by
+    # returning the first one in alphabetical order
+    def get_winner(self) -> str:
+        return max(self, key=self.get)
+
+    # Overriding the __repr__ method to include the winner
+    def __repr__(self):
+        return f"Winner: {self.winner}\n{super().__repr__()}"
+
+
+# Decorator to wrap the outcome functions, to accept both numpy arrays and
+# pandas dataframes and return a Result object
+def outcome_wrapper(func):
+    def wrapper(preferences: UnionArray, *args, **kwargs):
+        if isinstance(preferences, pd.DataFrame):
+            s = preferences.to_numpy()
+        else:
+            s = preferences
+
+        res = func(s, *args, **kwargs)
```

Comment:
Please add some explanations here as how the wrapper works and what actions it's performing on each voting scheme

Reply:
The wrapper checks the parameters of the wrapped function, converting a pandas df into a numpy array. It also converts the output of any wrapped function to the Result type